### PR TITLE
Granger causality

### PR DESCRIPTION
**Pull request by aoqiz**
**Update**

- Add auto test to Granger Causality. The ground truth `p_value_matrix_truth` and `adj_matrix_truth` are computed from the current version [b49980d](https://github.com/cmu-phil/causal-learn/commit/b49980d046607baaaa66ff8dc0ceb98452ab8616)

**Test plan**

`python -m unittest tests/TestGranger.py  # should pass`

image

**TODO**

- Add tests for more complicated data, such as non-Gaussian data, mix of continuous and discrete data.

### CODE REVIEW

File: tests/TestGranger.py [hunk 1]
```diff
@@ -54,8 +52,10 @@ def test_granger_test(self):
         dataset = self.syn_data_2d()
         G = Granger()
         p_value_matrix, adj_matrix = G.granger_test_2d(data=dataset)
-        print('P-value matrix is \n {}'.format(p_value_matrix))
-        print('Adjacency matrix is \n {}'.format(adj_matrix))
+        p_value_matrix_truth = np.array([[0, 0.5989, 0, 0.5397], [0.0006, 0, 0.0014, 0]])
+        adj_matrix_truth = np.array([[1, 0, 1, 0], [1, 1, 1, 1]])
```
Comment (tofuwen) [inline tests/TestGranger.py:56]: could you remind me how you get this truth_value?

File: tests/TestGranger.py [hunk 2]
```diff
@@ -66,10 +66,8 @@ def test_granger_lasso(self):
         dataset = self.syn_data_3d()
         G = Granger()
         coeff = G.granger_lasso(data=dataset)
-        print('Estimated matrix is \n {}'.format(coeff))
+        coeff_truth = np.array([[0.09, 0.1101, 0.1527, 0.1127, 0.0226, 0.1538],
```
Comment (tofuwen) [inline tests/TestGranger.py:69]: and what about this?

File: tests/TestGranger.py (hunk 1 above)
Comment (aoqiz) [reply tests/TestGranger.py:56]: It was computed from the current version [1ebf232](https://github.com/cmu-phil/causal-learn/commit/1ebf232f0244a60fa7908bdea2b1bdf451da4074)

File: tests/TestGranger.py (hunk 2 above)
Comment (aoqiz) [reply tests/TestGranger.py:69]: Same, computed from the current version [1ebf232](https://github.com/cmu-phil/causal-learn/commit/1ebf232f0244a60fa7908bdea2b1bdf451da4074).

File: tests/TestGranger.py (hunk 1 above)
Comment (tofuwen) [reply tests/TestGranger.py:56]: cool, please add comments on how the p-values are generated --- same as all other PRs (it's a good practice to check other PRs before writing the PR so that you know what are the best practices we follow currently :) ) This can reduce the review overhead.

File: tests/TestGranger.py (hunk 2 above)
Comment (tofuwen) [reply tests/TestGranger.py:69]: same, please add some comments :)

A good mental practice: write code in a way that, if you check the data a year later, do you know what the code means and how the values are generated?

**Review by tofuwen — COMMENTED** [review]
Please add comments, and I think this PR is good

**Comment by tofuwen** [comment]
And btw, did you check the p-values / coefficient produced by the codebase make sense? I remembered we figured out a bug by noticing p-values not making sense.

**Comment by tofuwen** [comment]
To add comments on how p-values are generated, here is an example: https://github.com/cmu-phil/causal-learn/pull/59/files

Something like: "# All the benchmark results of loaded files (e.g. "./TestData/benchmark_returned_results/")    #
# are obtained from the code of causal-learn as of commit                                      #
# https://github.com/cmu-phil/causal-learn/commit/129dcdf (07-14-2022).                        #
#                                                                                              #
# We are not sure if the results are completely "correct" (reflect ground truth graph) or not. #
# So if you find your tests failed, it means that your modified code is logically inconsistent #
# with the code as of 129dcdf, but not necessarily means that your code is "wrong".            #
# If you are sure that your modification is "correct" (e.g. fixed some bugs in 129dcdf),       #
# please report it to us. We will then modify these benchmark results accordingly. Thanks :)   #"

Basically the idea is: make sure others know how your p-values are generated (instead of the magic number in code) --- you just need to say we use code in commit xxxx to generate the p-values.

File: tests/TestGranger.py (hunk 1 above)
Comment (aoqiz) [reply tests/TestGranger.py:56]: Cool, thanks.

**Comment by aoqiz** [comment]
> And btw, did you check the p-values / coefficient produced by the codebase make sense? I remembered we figured out a bug by noticing p-values not making sense.

Yes, it make sense! Could you please report that bug? The current version works well to me.

**Comment by tofuwen** [comment]
> > And btw, did you check the p-values / coefficient produced by the codebase make sense? I remembered we figured out a bug by noticing p-values not making sense.
> 
> Yes, it make sense! Could you please report that bug? The current version works well to me.

Oh, I mean whether you checked the resulted p-values make sense or not. Previously we found out KCI p-values didn't make sense, that's why we figured out a bug in KCI, right?

**Comment by tofuwen** [comment]
After you check the p-values and think they make sense, then I can think we can push this PR.

**Comment by aoqiz** [comment]
> After you check the p-values and think they make sense, then I can think we can push this PR.

Yes, they make sense. I think we can push this PR. @tofuwen @kunwuz Could you please help push this PR? Thanks!

**Comment by tofuwen** [comment]
@kunwuz could you help merge this PR?
